aws/signer/v4/v4.go

```diff
@@ -0,0 +1,537 @@
+		ExpireTime:  0,
+		isPresign:   false,
+		ServiceName: service,
+		Region:      region,
+	}
+
+	if ctx.isRequestSigned() {
+		if !v4.Credentials.IsExpired() && time.Now().Before(ctx.Time.Add(10*time.Minute)) {
+			// If the request is already signed, and the credentials have not
+			// expired, and the request is not too old ignore the signing request.
+			return ctx.SignedHeaderVals, nil
+		}
+		ctx.Time = time.Now()
+	}
+
+	var err error
+	ctx.credValues, err = v4.Credentials.Get()
+	if err != nil {
+		return http.Header{}, err
+	}
+
+	if ctx.credValues.SessionToken != "" {
+		ctx.Request.Header.Set("X-Amz-Security-Token", ctx.credValues.SessionToken)
+	}
+
+	v4.build(ctx)
+
+	if v4.Debug.Matches(aws.LogDebugWithSigning) {
+		v4.logSigningInfo(ctx)
+	}
+
+	return ctx.SignedHeaderVals, nil
+}
+
+// Presign presigns AWS v4 requests
+func (v4 *Signer) Presign(r *http.Request, service, region string, exp time.Duration, signTime time.Time) (string, http.Header, error) {
+	buf, _ := ioutil.ReadAll(r.Body)
+	r.Body = ioutil.NopCloser(bytes.NewReader(buf))
+	return v4.presignWithBody(r, bytes.NewReader(buf), service, region, exp, signTime)
+}
+
+func (v4 *Signer) presignWithBody(r *http.Request, body io.ReadSeeker, service, region string, exp time.Duration, signTime time.Time) (string, http.Header, error) {
+	ctx := &signingCtx{
+		Request:    r,
+		Body:       body,
+		Query:      r.URL.Query(),
+		Time:       signTime,
+		ExpireTime: exp,
+		isPresign:  true,
+
+		ServiceName: service,
+		Region:      region,
+	}
+
+	if ctx.isRequestSigned() {
+		if !v4.Credentials.IsExpired() && time.Now().Before(ctx.Time.Add(10*time.Minute)) {
+			// If the request is already signed, and the credentials have not
+			// expired, and the request is not too old ignore the signing request.
+			return r.URL.String(), ctx.SignedHeaderVals, nil
+		}
+		ctx.Time = time.Now()
+
+		// The credentials have expired for this request. The current signing
+		// is invalid, and needs to be request because the request will fail.
+		ctx.removePresign()
+		// Update the request's query string to ensure the values stays in
+		// sync in the case retrieving the new credentials fails.
+		ctx.Request.URL.RawQuery = ctx.Query.Encode()
+	}
+
+	var err error
+	ctx.credValues, err = v4.Credentials.Get()
+	if err != nil {
+		return "", http.Header{}, err
+	}
+
+	ctx.Query.Set("X-Amz-Algorithm", authHeaderPrefix)
+	if ctx.credValues.SessionToken != "" {
+		ctx.Query.Set("X-Amz-Security-Token", ctx.credValues.SessionToken)
+	} else {
+		ctx.Query.Del("X-Amz-Security-Token")
+	}
+
+	v4.build(ctx)
+
+	if v4.Debug.Matches(aws.LogDebugWithSigning) {
+		v4.logSigningInfo(ctx)
+	}
+
+	return r.URL.String(), ctx.SignedHeaderVals, nil
+}
+
+// Sign requests with signature version 4.
+//
+// Will sign the requests with the service config's Credentials object
+// Signing is skipped if the credentials is the credentials.AnonymousCredentials
+// object.
+func Sign(req *request.Request) {
```

Comment:
Should include that this is a `request.Handler` to sign the requests made by the SDK's service clients in the comment for this function.